### PR TITLE
Tim bmdashboard entry point

### PR DESCRIPTION
**Pull request by tdkent**
# Description
Implement Phase 2 Entry Point: 
- URL to access the Building Management Dashboard
- Password protection to access the Dashboard

## Related PRS (if any):
To test this PR you need to checkout the [backend PR 543](https://github.com/OneCommunityGlobal/HGNRest/pull/543)

## Main changes explained:
- Adds URL entry point “/bmdashboard” to access Building Management Dashboard (ie Phase 2)
- Users must be logged in AND verified via login page at “bmdashboard/login” to access the BM Dashboard.

## How to test:
- Follow steps to test backend PR 543
- `git checkout Tim_bmdashboard_entry_point`
- `npm install`
- `npm run start:local`
- Log out of HGN and test that attempting to access “/bmdashboard” or “/bmdashboard/:whatever” redirects to “/login”
- Log into HGN as a volunteer user
- Test that attempting to access “/bmdashboard” or “/bmdashboard/:whatever” redirects to “bmdashboard/login”
- Check frontend validation of login form: 
   - Enter text that is not in correct email format (Note: a string with only two segments such as test@test is technically acceptable email format)
   - Enter an invalid password (< 8 characters)
- Check backend validation of login form: 
   - Enter an email for an account that does not exist, or is not the email you are signed in as
   - Enter the correct email but incorrect password
- Enter the correct email and password for the user you are already signed in as. Verify you are redirected to “/bmdashboard”.
- Also verify that attempting to access “bmdashboard/login” now redirects to “/bmdashboard”
- Finally, log out of HGN and verify you can no longer access “/bmdashboard”

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/98129080/90e6d0ca-88f8-4520-8d03-52bbf55ff1c1

## Note:
- There are no currently no role restrictions for access the BM Dashboard, so all testing can be done as a Volunteer user

### CODE REVIEW

**Review by Shiwani99 — APPROVED** [review]
Hi,
Great work. I tested your PR and it works as intended. 


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78906820/b8246b70-8693-4252-94ab-c8a265a0b6fe

**Review by DavidC0126 — APPROVED** [review]
Hi @tdkent I tested the PR and it works as intended. Good job!


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/74082492/9b822f98-8f40-49d9-9485-f28a7d56935d

**Review by wantingxu7 — APPROVED** [review]
Hi,
I tested and all functions work well. Good work!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105323719/e8a375aa-8d4f-4831-944b-e1583a048dec

**Comment by astrjc0326** [comment]
Hello,
I tested the functions as mentioned. It works perfectly. Good Job!


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/98345990/23e63d33-e757-4821-abea-7e0011d496bd

**Review by navneeeth — APPROVED** [review]
Hey @tdkent! Tested this PR along with the related backend PR and the changes work as expected. Good implementation!

During my first test, I encountered errors related to getting access to the bmdashboard before logging into the bmdashboard while the /bmdashboard/login route still redirected me to the login page:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/db0f435a-638f-4e0c-8b9f-900348386312


Once I removed caches and reloaded the page and also re-compiled and ran the backend and frontend, the PR worked as intended:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44264834/a16e40dd-6396-41c5-8980-84794708e983

**Review by Jiyuanxia — APPROVED** [review]
Hi @tdkent , I have reviewed your pull request and tested it. All the redirection functions and validations works as you intended. Good job!
Frontend email check:
Screenshot 2023-09-21 174229
Frontend password check:
Screenshot 2023-09-21 173657
Backend email check:
Screenshot 2023-09-21 173801
Backend password check:
Screenshot 2023-09-21 173737
Redirection successful:
Screenshot 2023-09-21 173823

**Review by Haoj1 — APPROVED** [review]
@tdkent Good job. The new authentication login looks good to me!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/99279909/def286c7-cd2f-4375-bd47-a58471bae4fa

**Review by JYXiao-2021 — APPROVED** [review]
Hi, I test the pr with backend and frontend and it works as described. Great job.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/49748716/b334a424-e9c3-4e28-b935-f6af5ff2af25

**Review by JeffLi117 — APPROVED** [review]
Changes work as intended. 

After logging in and trying to access "bmdashboard/", I arrive at "bmdashboard/login" and test the email and password, respectively: 

image
image

Upon entering the correct login information, I am redirected to “/bmdashboard”:
image

Once I log out, I am unable to access “/bmdashboard”.

**Review by tsunami776 — APPROVED** [review]
Hi @tdkent, I followed your instructions for testing bmdashboard, and it worked as described in your PR. Great work!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/43768723/7eca6f68-29f4-4d46-b95b-3874892eda2e

**Review by TuanDinh140194 — APPROVED** [review]
Hi! I test the PR and everything is look good as describe. Very good job !

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/81892917/42ee46e6-ca9b-45fc-894c-8a9c34588938

File: src/components/BMDashboard/Login/BMLogin.jsx [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import { useEffect, useState } from 'react';
+import { connect } from 'react-redux';
+import { Redirect, withRouter } from 'react-router-dom';
+import { Form, FormGroup, FormText, Input, Label, Button, FormFeedback } from 'reactstrap';
+import Joi from 'joi';
+
+import { loginBMUser } from 'actions/authActions';
+
+const BMLogin = (props) => {
```
Comment (TuanDinh140194) [inline src/components/BMDashboard/Login/BMLogin.jsx:9]: You are already destructuring props at the beginning of your component. However, you can further improve it by destructuring individual props directly in the function parameters:
const BMLogin = ({ dispatch, auth, history, location }) => {
  // ...
}

**Comment by one-community** [comment]
Thank you everyone, merging!

**Comment by Vishala09** [comment]
Works Perfect! Awesome Work!

[Untitled_ Sep 26,2023 3_33 PM.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/65274029/218fbcd0-e586-470c-bdbb-5b4098d87579)

**Review by Vishala09 — COMMENTED** [review]
Works Perfect! Awesome Work!

**Pull request merged**